### ISSUE TITLE
[BUG] Wrong public key used to encrypt card

### ISSUE THREAD

**Issue by mcarter6061** [Adyen/adyen-ios#797]
**Describe the bug**
We have multiple client keys within our app, one per merchant account.  When making payments to several merchant accounts in a single app session with a non-stored card, we see a success on all calls to the first merchant account used and "422: Unable to decrypt data" on payments to the other merchant accounts.

Looking through the SDK code it seems a public key is fetched to encrypt the card data for a particular client key.  If you make a second payment which would use another client key the first public key is used rather than fetching a new public key, as it is stored as a static on the `cardPublicKeyProvider`:
```
    internal static var cachedCardPublicKey: String?
```

**To Reproduce**
With a client configured with multiple client keys:

1. Initiate a payment against first client key, using a non stored card.
2. Initiate a second payment for a different client key, using a non stored card.

**Expected behavior**
The first payment and the second payment should both succeed (though for this test just sending the payment data is enough, there could be a redirect and the user can abandon the payment ).

Actual behaviour, first payment succeeds and second payment fails with the 422: Unable to decrypt data error.

**Smartphone (please complete the following information):**
 - Device: iPhone 12 Pro (simulator)
 - OS: iOS 15.0 (simulator)
 - SDK Version 4.3.0

**Additional context**
I see develop has removed the `CardPublicKeyProvider` class however it looks like the `PublicKeyProvider` class has the same behaviour.

**Comment by mohammedDehairy** [comment]
Hi @mcarter6061 ,

Thanks for reporting this, the `cardComponent` indeed assumes one clientKey/MerchantAccount per app session, we can definitely fix this and allow for multiple clientKey(s)/MerchantAccount(s) per app session.

I am just curious, can you explain your use case? why would you need to have multiple clientKey(s)/MerchantAccount(s) per app?.

Thanks

**Comment by mcarter6061** [comment]
I was not involved with the Adyen configuration I'm afraid, so I'm not sure why we have this setup.  We have two types of payment journeys in the app ( one for purchasing products and one for purchasing subscriptions), and each journey has a different merchant account and client key.  I'll ask and see if I can find out why it is setup this way.  

@mohammedDehairy Is it possible to configure one client key for multiple merchant accounts?

**Comment by mohammedDehairy** [comment]
@mcarter6061 

```
Is it possible to configure one client key for multiple merchant accounts?
```

Yes, a clientKey is for each web service user and a web service user is on the company level, and can be used for multiple merchant accounts.

**Comment by mcarter6061** [comment]
What I've worked out so far is that the app calls different endpoints for payment based of which journey they are in.  Those endpoints are configured with different web service users (and therefore API keys / client keys ) when making the `/payments` call to Adyen.

@mohammedDehairy I think the client key is owned by the web service user, so as long as our backend services have multiple users we will have to have multiple client keys?  Does that sound correct and is this a configuration you would not expect?

**Comment by mohammedDehairy** [comment]
@mcarter6061 

```
Does that sound correct?
```
yes that's correct

```
 and is this a configuration you would not expect?
```
its a bit surprising to us tbh :-), I mean this code has been live for quite a long time and no merchant complained, so this use case is a bit an outlier, but never the less, its a good point and I already fixed this issue in this [pull request](https://github.com/Adyen/adyen-ios/pull/798), can you have a look and let me know if it works for you?.

**Comment by descorp** [comment]
Hey @mcarter6061 

The fix was released in 4.7.1
I am going to close the ticket, since the issue is confirmed to be fixed.
